File: src/net/minecraft/client/render/entity/LivingEntityRenderer.cpp
#include "net/minecraft/client/render/entity/LivingEntityRenderer.hpp"
#include "net/minecraft/client/gl/GlConstants.hpp"
#include <cmath>
#include "net/minecraft/client/font/TextRenderer.hpp"
#include "net/minecraft/client/render/Tessellator.hpp"
#include "net/minecraft/client/render/entity/EntityRenderDispatcher.hpp"
#include "net/minecraft/entity/Entity.hpp"
#include "net/minecraft/entity/LivingEntity.hpp"
#include "net/minecraft/entity/mob/CreeperEntity.hpp"
#include "net/minecraft/entity/mob/GhastEntity.hpp"
#include "net/minecraft/entity/mob/GiantEntity.hpp"
#include "net/minecraft/entity/mob/PigZombieEntity.hpp"
#include "net/minecraft/entity/mob/SkeletonEntity.hpp"
#include "net/minecraft/entity/mob/SlimeEntity.hpp"
#include "net/minecraft/entity/mob/SpiderEntity.hpp"
#include "net/minecraft/entity/mob/ZombieEntity.hpp"
#include "net/minecraft/entity/passive/ChickenEntity.hpp"
#include "net/minecraft/entity/passive/CowEntity.hpp"
#include "net/minecraft/entity/passive/PigEntity.hpp"
#include "net/minecraft/entity/passive/SheepEntity.hpp"
#include "net/minecraft/entity/passive/SquidEntity.hpp"
#include "net/minecraft/entity/passive/WolfEntity.hpp"
#include "net/minecraft/entity/player/PlayerEntity.hpp"
#include "net/minecraft/mod/runtime/LuaDirectHooks.hpp"
#include "net/minecraft/mod/runtime/LuaEntityBindings.hpp"
#include "net/minecraft/util/math/MathHelper.hpp"
#include "net/minecraft/client/render/RenderSystem.hpp"
#include "net/minecraft/client/render/RenderType.hpp"
namespace net::minecraft::client::render::entity {
namespace {
std::string_view entityTypeName(const net::minecraft::entity::LivingEntity& entity) {
 using namespace net::minecraft::entity;
 if(dynamic_cast<const player::PlayerEntity*>(&entity)) {
  return "player";
 }
 if(dynamic_cast<const mob::ZombieEntity*>(&entity)) {
  return "zombie";
 }
 if(dynamic_cast<const mob::SkeletonEntity*>(&entity)) {
  return "skeleton";
 }
 if(dynamic_cast<const mob::CreeperEntity*>(&entity)) {
  return "creeper";
 }
 if(dynamic_cast<const mob::PigZombieEntity*>(&entity)) {
  return "pigzombie";
 }
 if(dynamic_cast<const mob::SpiderEntity*>(&entity)) {
  return "spider";
 }
 if(dynamic_cast<const mob::SlimeEntity*>(&entity)) {
  return "slime";
 }
 if(dynamic_cast<const mob::GhastEntity*>(&entity)) {
  return "ghast";
 }
 if(dynamic_cast<const mob::GiantEntity*>(&entity)) {
  return "giant";
 }
 if(dynamic_cast<const passive::PigEntity*>(&entity)) {
  return "pig";
 }
 if(dynamic_cast<const passive::CowEntity*>(&entity)) {
  return "cow";
 }
 if(dynamic_cast<const passive::ChickenEntity*>(&entity)) {
  return "chicken";
 }
 if(dynamic_cast<const passive::SheepEntity*>(&entity)) {
  return "sheep";
 }
 if(dynamic_cast<const passive::WolfEntity*>(&entity)) {
  return "wolf";
 }
 if(dynamic_cast<const passive::SquidEntity*>(&entity)) {
  return "squid";
 }
 return "living";
}
} // namespace
LivingEntityRenderer::LivingEntityRenderer(model::EntityModel* modelIn, float shadowRadiusIn) : model(modelIn) {
 shadowRadius = shadowRadiusIn;
}
void LivingEntityRenderer::setDecorationModel(model::EntityModel* modelIn) {
 decorationModel = modelIn;
}
void LivingEntityRenderer::render(
    const net::minecraft::Entity& entity, double x, double y, double z, float /*yaw*/, float tickDelta) {
 const auto* living = dynamic_cast<const net::minecraft::LivingEntity*>(&entity);
 if(living == nullptr || model == nullptr) {
  return;
 }
 render::RenderPassScope passScope(render::RenderType::entityCutout());
 { // entityMatrix must be popped before renderNameTag, which pushes its own unrelated matrix.
  RenderSystem::pushMatrix();
  RenderSystem::disableCull();
  model->handSwingProgress = getHandSwingProgress(*living, tickDelta);
  if(decorationModel != nullptr) {
   decorationModel->handSwingProgress = model->handSwingProgress;
  }
  model->riding = living->hasVehicle();
  if(decorationModel != nullptr) {
   decorationModel->riding = model->riding;
  }
  const float baseBodyYaw = living->lastBodyYaw + (living->bodyYaw - living->lastBodyYaw) * tickDelta;
  const float baseHeadYaw = living->prevYaw + (living->yaw - living->prevYaw) * tickDelta;
  const float baseHeadPitch = living->prevPitch + (living->pitch - living->prevPitch) * tickDelta;
  float limbAngle =
      living->lastWalkAnimationSpeed + (living->walkAnimationSpeed - living->lastWalkAnimationSpeed) * tickDelta;
  float limbDistance = living->walkAnimationProgress - living->walkAnimationSpeed * (1.0f - tickDelta);
  if(limbAngle > 1.0f) {
   limbAngle = 1.0f;
  }
  float bodyYaw = baseBodyYaw;
  float headYawRel = baseHeadYaw - baseBodyYaw;
  float headPitch = baseHeadPitch;
  float poseScale = 1.0f;
  float offsetX = 0.0f;
  float offsetY = 0.0f;
  float offsetZ = 0.0f;
  model->poseActive = false;
  model->partOverrides.clear();
  if(decorationModel != nullptr) {
   decorationModel->poseActive = false;
   decorationModel->partOverrides.clear();
  }
  net::minecraft::mod::EntityRenderPose pose{};
  pose.bodyYaw = bodyYaw;
  pose.headYaw = headYawRel;
  pose.headPitch = headPitch;
  pose.limbSwing = limbAngle;
  pose.limbDistance = limbDistance;
  net::minecraft::mod::runtime::applyRegisteredPoseHooks(*living, tickDelta, pose);
  if(net::minecraft::mod::runtime::hasLuaHook(net::minecraft::mod::runtime::LuaEventId::EntityRender)) {
    net::minecraft::mod::EntityRenderEvent event;
    event.entity = living;
    event.entityId = living->id;
    event.entityType = std::string(entityTypeName(*living));
    event.isPlayer = dynamic_cast<const net::minecraft::entity::player::PlayerEntity*>(living) != nullptr;
    event.tickDelta = tickDelta;
    event.pose = pose;
    net::minecraft::mod::runtime::luaHookEntityRender(event);
   pose = event.pose;
  }
  bodyYaw = pose.bodyYaw;
  headYawRel = pose.headYaw;
  headPitch = pose.headPitch;
  limbAngle = pose.limbSwing;
  limbDistance = pose.limbDistance;
  poseScale = pose.scale;
  offsetX = pose.offsetX;
  offsetY = pose.offsetY;
  offsetZ = pose.offsetZ;
  if(!pose.parts.empty()) {
   model->poseActive = true;
   constexpr float kDegToRad = 0.01745329251994329547f;
   for(const auto& [name, part] : pose.parts) {
    auto& out = model->partOverrides[name];
    out.yaw = std::isnan(part.yaw) ? part.yaw : part.yaw * kDegToRad;
    out.pitch = std::isnan(part.pitch) ? part.pitch : part.pitch * kDegToRad;
    out.roll = std::isnan(part.roll) ? part.roll : part.roll * kDegToRad;
   }
  }
  applyTranslation(*living, x, y, z);
  if(offsetX != 0.0f || offsetY != 0.0f || offsetZ != 0.0f) {
   RenderSystem::translate(offsetX, offsetY, offsetZ);
  }
  const float headBob = getHeadBob(*living, tickDelta);
  applyHandSwingRotation(*living, headBob, bodyYaw, tickDelta);
  constexpr float scaleUnit = 0.0625f;
  RenderSystem::scale(-1.0f, -1.0f, 1.0f);
  applyScale(*living, tickDelta);
  if(poseScale != 1.0f) {
   RenderSystem::scale(poseScale, poseScale, poseScale);
  }
  RenderSystem::translate(0.0f, -24.0f * scaleUnit - 0.0078125f, 0.0f);
  (void)bindDownloadedTexture(living->skinUrl, living->getTexture());
  model->animateModel(const_cast<net::minecraft::LivingEntity&>(*living), limbDistance, limbAngle, tickDelta);
  model->render(limbDistance, limbAngle, headBob, headYawRel, headPitch, scaleUnit);
  auto syncDecorationPose = [&]() {
   if(decorationModel != nullptr) {
    decorationModel->poseActive = model->poseActive;
    decorationModel->partOverrides = model->partOverrides;
   }
  };
  for(int layer = 0; layer < 4; ++layer) {
   if(!bindTexture(*living, layer, tickDelta)) {
    continue;
   }
   if(decorationModel != nullptr) {
    syncDecorationPose();
    decorationModel->render(limbDistance, limbAngle, headBob, headYawRel, headPitch, scaleUnit);
   }
   RenderSystem::disableBlend();
  }
  renderMore(*living, tickDelta);
  const float brightness = living->getBrightnessAtEyes(tickDelta);
  const int overlay = getOverlayColor(*living, brightness, tickDelta);
  if(((overlay >> 24) & 0xFF) > 0 || living->hurtTime > 0 || living->deathTime > 0) {
   render::RenderPassScope overlayScope(render::RenderType::entityCutout());
   RenderSystem::disableCull();
   RenderSystem::disableTexture();
   RenderSystem::enableBlend();
   RenderSystem::blendFunc(0x0302, 0x0303); // GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA
   RenderSystem::depthFunc(0x0202); // GL_EQUAL
   if(living->hurtTime > 0 || living->deathTime > 0) {
    RenderSystem::color4f(brightness, 0.0f, 0.0f, 0.4f);
    model->render(limbDistance, limbAngle, headBob, headYawRel, headPitch, scaleUnit);
    for(int layer = 0; layer < 4; ++layer) {
     if(!bindDecorationTexture(*living, layer, tickDelta)) {
      continue;
     }
     RenderSystem::color4f(brightness, 0.0f, 0.0f, 0.4f);
     if(decorationModel != nullptr) {
      syncDecorationPose();
      decorationModel->render(limbDistance, limbAngle, headBob, headYawRel, headPitch, scaleUnit);
     }
    }
   }
   if(((overlay >> 24) & 0xFF) > 0) {
    const float cr = static_cast<float>((overlay >> 16) & 0xFF) / 255.0f;
    const float cg = static_cast<float>((overlay >> 8) & 0xFF) / 255.0f;
    const float cb = static_cast<float>(overlay & 0xFF) / 255.0f;
    const float ca = static_cast<float>((overlay >> 24) & 0xFF) / 255.0f;
    RenderSystem::color4f(cr, cg, cb, ca);
    model->render(limbDistance, limbAngle, headBob, headYawRel, headPitch, scaleUnit);
    for(int layer = 0; layer < 4; ++layer) {
     if(!bindDecorationTexture(*living, layer, tickDelta)) {
      continue;
     }
     RenderSystem::color4f(cr, cg, cb, ca);
     if(decorationModel != nullptr) {
      syncDecorationPose();
      decorationModel->render(limbDistance, limbAngle, headBob, headYawRel, headPitch, scaleUnit);
     }
    }
   }
   RenderSystem::depthTest();
   RenderSystem::disableBlend();
   RenderSystem::enableTexture();
  }
  RenderSystem::cullBackFaces();
  RenderSystem::popMatrix();
 } // pop entityMatrix here, before renderNameTag pushes its own matrix
 renderNameTag(*living, x, y, z);
}
void LivingEntityRenderer::applyTranslation(const net::minecraft::LivingEntity& entity, double x, double y, double z) {
 (void)entity;
 RenderSystem::translate(static_cast<float>(x), static_cast<float>(y), static_cast<float>(z));
}
void LivingEntityRenderer::applyHandSwingRotation(const net::minecraft::LivingEntity& entity,
                                                  float /*headBob*/,
                                                  float bodyYaw,
                                                  float tickDelta) {
 RenderSystem::rotate(180.0f - bodyYaw, 0.0f, 1.0f, 0.0f);
 if(entity.deathTime > 0) {
  float deathAnim = (static_cast<float>(entity.deathTime) + tickDelta - 1.0f) / 20.0f * 1.6f;
  deathAnim = MathHelper::sqrt(deathAnim);
  if(deathAnim > 1.0f) {
   deathAnim = 1.0f;
  }
  RenderSystem::rotate(deathAnim * getDeathYaw(entity), 0.0f, 0.0f, 1.0f);
 }
}
float LivingEntityRenderer::getHandSwingProgress(const net::minecraft::LivingEntity& entity, float tickDelta) const {
 return entity.getHandSwingProgress(tickDelta);
}
float LivingEntityRenderer::getHeadBob(const net::minecraft::LivingEntity& entity, float tickDelta) const {
 return static_cast<float>(entity.age) + tickDelta;
}
void LivingEntityRenderer::renderMore(const net::minecraft::LivingEntity& entity, float tickDelta) {
 (void)entity;
 (void)tickDelta;
}
bool LivingEntityRenderer::bindDecorationTexture(const net::minecraft::LivingEntity& entity,
                                                 int layer,
                                                 float tickDelta) {
 return bindTexture(entity, layer, tickDelta);
}
bool LivingEntityRenderer::bindTexture(const net::minecraft::LivingEntity& entity, int layer, float tickDelta) {
 (void)entity;
 (void)layer;
 (void)tickDelta;
 return false;
}
float LivingEntityRenderer::getDeathYaw(const net::minecraft::LivingEntity& entity) const {
 (void)entity;
 return 90.0f;
}
int LivingEntityRenderer::getOverlayColor(const net::minecraft::LivingEntity& entity,
                                          float brightness,
                                          float tickDelta) const {
 (void)entity;
 (void)brightness;
 (void)tickDelta;
 return 0;
}
void LivingEntityRenderer::applyScale(const net::minecraft::LivingEntity& entity, float tickDelta) {
 (void)entity;
 (void)tickDelta;
}
void LivingEntityRenderer::renderNameTag(const net::minecraft::LivingEntity& entity, double x, double y, double z) {
 if(dispatcher != nullptr && dispatcher->options().debugHud) {
  renderNameTag(entity, std::to_string(entity.id), x, y, z, 64);
 }
}
void LivingEntityRenderer::renderNameTag(
    const net::minecraft::LivingEntity& entity, const std::string& name, double x, double y, double z, int range) {
 if(dispatcher == nullptr || dispatcher->cameraEntity() == nullptr) {
  return;
 }
 const float distance = entity.getDistance(*dispatcher->cameraEntity());
 if(distance > static_cast<float>(range)) {
  return;
 }
 font::TextRenderer* textRenderer = getTextRenderer();
 if(textRenderer == nullptr) {
  return;
 }
 constexpr float nameScale = 1.6f;
 const float pixelScale = 0.016666668f * nameScale;
 {
  RenderSystem::pushMatrix();
  RenderSystem::translate(static_cast<float>(x), static_cast<float>(y) + 2.3f, static_cast<float>(z));
  RenderSystem::rotate(-dispatcher->yaw_, 0.0f, 1.0f, 0.0f);
  RenderSystem::rotate(dispatcher->pitch_, 1.0f, 0.0f, 0.0f);
  RenderSystem::scale(-pixelScale, -pixelScale, pixelScale);
  render::RenderPassScope passScope(render::RenderType::guiTextured());
  Tessellator& tessellator = Tessellator::INSTANCE;
  int yOffset = 0;
  if(name == "deadmau5") {
   yOffset = -10;
  }
  RenderSystem::disableTexture();
  tessellator.startQuads();
  const int halfWidth = textRenderer->getWidth(name) / 2;
  tessellator.color(0.0f, 0.0f, 0.0f, 0.25f);
  tessellator.vertex(-halfWidth - 1, -1 + yOffset, 0.0);
  tessellator.vertex(-halfWidth - 1, 8 + yOffset, 0.0);
  tessellator.vertex(halfWidth + 1, 8 + yOffset, 0.0);
  tessellator.vertex(halfWidth + 1, -1 + yOffset, 0.0);
  tessellator.draw();
  {
   RenderSystem::enableTexture();
   textRenderer->draw(name, -textRenderer->getWidth(name) / 2, yOffset, 0x20FFFFFF);
   RenderSystem::depthTest();
   RenderSystem::depthMask(true);
   textRenderer->draw(name, -textRenderer->getWidth(name) / 2, yOffset, -1);
  }
  RenderSystem::popMatrix();
 }
}
} // namespace net::minecraft::client::render::entity

File: mcp/src/net/minecraft/client/render/entity/LivingEntityRenderer.java
/*
 * Decompiled with CFR 0.1.1 (FabricMC 57d88659).
 * 
 * Could not load the following classes:
 *  net.fabricmc.api.EnvType
 *  net.fabricmc.api.Environment
 */
package net.minecraft.client.render.entity;

import net.fabricmc.api.EnvType;
import net.fabricmc.api.Environment;
import net.minecraft.client.Minecraft;
import net.minecraft.client.font.TextRenderer;
import net.minecraft.client.render.Tessellator;
import net.minecraft.client.render.entity.EntityRenderer;
import net.minecraft.client.render.entity.model.EntityModel;
import net.minecraft.entity.LivingEntity;
import net.minecraft.util.math.MathHelper;
import org.lwjgl.opengl.GL11;

@Environment(value=EnvType.CLIENT)
public class LivingEntityRenderer
extends EntityRenderer {
    protected EntityModel model;
    protected EntityModel decorationModel;

    public LivingEntityRenderer(EntityModel entityModel, float shadowRadius) {
        this.model = entityModel;
        this.shadowRadius = shadowRadius;
    }

    public void setDecorationModel(EntityModel model) {
        this.decorationModel = model;
    }

    public void render(LivingEntity arg, double d, double e, double f, float g, float h) {
        GL11.glPushMatrix();
        GL11.glDisable(2884);
        this.model.handSwingProgress = this.getHandSwingProgress(arg, h);
        if (this.decorationModel != null) {
            this.decorationModel.handSwingProgress = this.model.handSwingProgress;
        }
        this.model.riding = arg.hasVehicle();
        if (this.decorationModel != null) {
            this.decorationModel.riding = this.model.riding;
        }
        try {
            float f2 = arg.lastBodyYaw + (arg.bodyYaw - arg.lastBodyYaw) * h;
            float f3 = arg.prevYaw + (arg.yaw - arg.prevYaw) * h;
            float f4 = arg.prevPitch + (arg.pitch - arg.prevPitch) * h;
            this.applyTranslation(arg, d, e, f);
            float f5 = this.getHeadBob(arg, h);
            this.applyHandSwingRotation(arg, f5, f2, h);
            float f6 = 0.0625f;
            GL11.glEnable(32826);
            GL11.glScalef(-1.0f, -1.0f, 1.0f);
            this.applyScale(arg, h);
            GL11.glTranslatef(0.0f, -24.0f * f6 - 0.0078125f, 0.0f);
            float f7 = arg.lastWalkAnimationSpeed + (arg.walkAnimationSpeed - arg.lastWalkAnimationSpeed) * h;
            float f8 = arg.walkAnimationProgress - arg.walkAnimationSpeed * (1.0f - h);
            if (f7 > 1.0f) {
                f7 = 1.0f;
            }
            this.bindDownloadedTexture(arg.skinUrl, arg.getTexture());
            GL11.glEnable(3008);
            this.model.animateModel(arg, f8, f7, h);
            this.model.render(f8, f7, f5, f3 - f2, f4, f6);
            for (int i = 0; i < 4; ++i) {
                if (!this.bindTexture(arg, i, h)) continue;
                this.decorationModel.render(f8, f7, f5, f3 - f2, f4, f6);
                GL11.glDisable(3042);
                GL11.glEnable(3008);
            }
            this.renderMore(arg, h);
            float f9 = arg.getBrightnessAtEyes(h);
            int n = this.getOverlayColor(arg, f9, h);
            if ((n >> 24 & 0xFF) > 0 || arg.hurtTime > 0 || arg.deathTime > 0) {
                GL11.glDisable(3553);
                GL11.glDisable(3008);
                GL11.glEnable(3042);
                GL11.glBlendFunc(770, 771);
                GL11.glDepthFunc(514);
                if (arg.hurtTime > 0 || arg.deathTime > 0) {
                    GL11.glColor4f(f9, 0.0f, 0.0f, 0.4f);
                    this.model.render(f8, f7, f5, f3 - f2, f4, f6);
                    for (int i = 0; i < 4; ++i) {
                        if (!this.bindDecorationTexture(arg, i, h)) continue;
                        GL11.glColor4f(f9, 0.0f, 0.0f, 0.4f);
                        this.decorationModel.render(f8, f7, f5, f3 - f2, f4, f6);
                    }
                }
                if ((n >> 24 & 0xFF) > 0) {
                    float f10 = (float)(n >> 16 & 0xFF) / 255.0f;
                    float f11 = (float)(n >> 8 & 0xFF) / 255.0f;
                    float f12 = (float)(n & 0xFF) / 255.0f;
                    float f13 = (float)(n >> 24 & 0xFF) / 255.0f;
                    GL11.glColor4f(f10, f11, f12, f13);
                    this.model.render(f8, f7, f5, f3 - f2, f4, f6);
                    for (int i = 0; i < 4; ++i) {
                        if (!this.bindDecorationTexture(arg, i, h)) continue;
                        GL11.glColor4f(f10, f11, f12, f13);
                        this.decorationModel.render(f8, f7, f5, f3 - f2, f4, f6);
                    }
                }
                GL11.glDepthFunc(515);
                GL11.glDisable(3042);
                GL11.glEnable(3008);
                GL11.glEnable(3553);
            }
            GL11.glDisable(32826);
        }
        catch (Exception exception) {
            exception.printStackTrace();
        }
        GL11.glEnable(2884);
        GL11.glPopMatrix();
        this.renderNameTag(arg, d, e, f);
    }

    protected void applyTranslation(LivingEntity entity, double headBob, double bodyYaw, double tickDelta) {
        GL11.glTranslatef((float)headBob, (float)bodyYaw, (float)tickDelta);
    }

    protected void applyHandSwingRotation(LivingEntity entity, float dx, float dy, float dz) {
        GL11.glRotatef(180.0f - dy, 0.0f, 1.0f, 0.0f);
        if (entity.deathTime > 0) {
            float f = ((float)entity.deathTime + dz - 1.0f) / 20.0f * 1.6f;
            if ((f = MathHelper.sqrt(f)) > 1.0f) {
                f = 1.0f;
            }
            GL11.glRotatef(f * this.getDeathYaw(entity), 0.0f, 0.0f, 1.0f);
        }
    }

    protected float getHandSwingProgress(LivingEntity entity, float tickDelta) {
        return entity.getHandSwingProgress(tickDelta);
    }

    protected float getHeadBob(LivingEntity entity, float tickDelta) {
        return (float)entity.age + tickDelta;
    }

    protected void renderMore(LivingEntity entity, float tickDelta) {
    }

    protected boolean bindDecorationTexture(LivingEntity entity, int i, float f) {
        return this.bindTexture(entity, i, f);
    }

    protected boolean bindTexture(LivingEntity mob, int layer, float tickDelta) {
        return false;
    }

    protected float getDeathYaw(LivingEntity entity) {
        return 90.0f;
    }

    protected int getOverlayColor(LivingEntity entity, float brightness, float timeDelta) {
        return 0;
    }

    protected void applyScale(LivingEntity entity, float scale) {
    }

    protected void renderNameTag(LivingEntity entity, double dx, double dy, double dz) {
        if (Minecraft.isDebugProfilerEnabled()) {
            this.renderNameTag(entity, Integer.toString(entity.id), dx, dy, dz, 64);
        }
    }

    protected void renderNameTag(LivingEntity entity, String name, double dx, double dy, double dz, int range) {
        float f = entity.getDistance(this.dispatcher.cameraEntity);
        if (f > (float)range) {
            return;
        }
        TextRenderer textRenderer = this.getTextRenderer();
        float f2 = 1.6f;
        float f3 = 0.016666668f * f2;
        GL11.glPushMatrix();
        GL11.glTranslatef((float)dx + 0.0f, (float)dy + 2.3f, (float)dz);
        GL11.glNormal3f(0.0f, 1.0f, 0.0f);
        GL11.glRotatef(-this.dispatcher.yaw, 0.0f, 1.0f, 0.0f);
        GL11.glRotatef(this.dispatcher.pitch, 1.0f, 0.0f, 0.0f);
        GL11.glScalef(-f3, -f3, f3);
        GL11.glDisable(2896);
        GL11.glDepthMask(false);
        GL11.glDisable(2929);
        GL11.glEnable(3042);
        GL11.glBlendFunc(770, 771);
        Tessellator tessellator = Tessellator.INSTANCE;
        int n = 0;
        if (name.equals("deadmau5")) {
            n = -10;
        }
        GL11.glDisable(3553);
        tessellator.startQuads();
        int n2 = textRenderer.getWidth(name) / 2;
        tessellator.color(0.0f, 0.0f, 0.0f, 0.25f);
        tessellator.vertex(-n2 - 1, -1 + n, 0.0);
        tessellator.vertex(-n2 - 1, 8 + n, 0.0);
        tessellator.vertex(n2 + 1, 8 + n, 0.0);
        tessellator.vertex(n2 + 1, -1 + n, 0.0);
        tessellator.draw();
        GL11.glEnable(3553);
        textRenderer.draw(name, -textRenderer.getWidth(name) / 2, n, 0x20FFFFFF);
        GL11.glEnable(2929);
        GL11.glDepthMask(true);
        textRenderer.draw(name, -textRenderer.getWidth(name) / 2, n, -1);
        GL11.glEnable(2896);
        GL11.glDisable(3042);
        GL11.glColor4f(1.0f, 1.0f, 1.0f, 1.0f);
        GL11.glPopMatrix();
    }
}

